engine/geometry/SpatialCubicSpline.cs

```diff
@@ -65,5 +83,67 @@ public override Vector3 GetEndPosition()
 		{
 			return new Vector3(X.GetValueAt(1.0f), Y.GetValueAt(1.0f), Z.GetValueAt(1.0f));
 		}
+		
+		private void GenerateCurveNormal()
+		{
+			// Generate a tangent at key points:
+			// TODO: make the precision adapt to the precision of centerCurve!
+			
+			SortedList<float, float> normalX = new SortedList<float, float>();
+			SortedList<float, float> normalY = new SortedList<float, float>();
+			SortedList<float, float> normalZ = new SortedList<float, float>();
+			
+			// No normal supplied, pick arbitrary normal vector:
+			Vector3 up = new Vector3(0.0f, 0.0f, 1.0f);
+			if (Vector3.Dot(GetTangentAt(0.0f), new Vector3(0.0f, 0.0f, 1.0f)) < 0.1)
+			{
+				up = new Vector3(0.0f, 1.0f, 0.0f);
+			}
+			
+			// Traverse through all the points in the spline, and calculate the next normal, maintaining smoothness by
+			// memorizing the previous normal:
+			
+			float previousValue = X.PointsX[0]; // X.PointsX is guaranteed to have at least 2 entries, this is safe.
+			for (int i = 1; i < X.PointsX.Count; i++)
+			{
+				float currentValue = X.PointsX[i];
+				float segmentSize = currentValue - previousValue;
+				
+				// Interpolate between each two points:
+				
+				int stepCount = 4; // Arbitrary number of interpolating steps. Four should be enough to prevent sudden jumps.
+				for (int j = 0; j < stepCount; j++)
+				{
+					float t = previousValue + segmentSize*(float)j/(float)stepCount;
```

Comment:
```suggestion
					float t = (float)System.Math.FusedMultiplyAdd(segmentSize, (float)j/(float)stepCount, previousValue);
```

FusedMultiplyAdd is the most precision-preserving way to do a "multiply two values and then add a third" operation. I spotted it here but you can almost certainly find other spots to do it if you go hunting for them.

Reply:
Ugh, turns out FusedMultiplyAdd doesn't exist until .NET 5.0 and we're targeting 4.7.2. Never mind.

Reply:
Hmmm, that's annoying. FusedMultiplyAdd is quite a nice function indeed. Do you happen to know if we can upgrade to .NET 5.0? Or are we stuck with 4.7.2 for the time being?